workflows/rnaseq-ref-index/make_pre_mrna_fasta.R

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env Rscript
+
+# This script takes the input reference fasta of the genome and the 
+# corresponding gtf file, identifies the regions of interest corresponding
+# to spliced transcripts and unspliced transcripts, then subsets the genome 
+# and gtf for only those particular regions, and finally outputs an expanded.fa, 
+# expanded.gtf, and expanded.tx2gene.tsv all corresponding to the genomic 
+# regions with spliced and unspliced transcripts. 
+
+# load needed packages
+library(magrittr)
+library(optparse)
+library(eisaR)
+library(Biostrings)
+library(GenomicFeatures)
+library(BSgenome)
+
+# Set up optparse options
+option_list <- list(
+  make_option(
+    opt_str = c("-f", "--gtf"),
+    type = "character",
+    default = "annotation/Homo_sapiens.GRCh38.103.gtf.gz",
+    help = "File path for input gtf file",
+  ),
+  make_option(
+    opt_str = c("-g", "--genome"),
+    type = "character",
+    default = "fasta/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz",
+    help = "File path for reference fasta file",
+  ), 
+  make_option(
+    opt_str = c("-O", "--output_dir"),
+    type = "character",
+    default = "expanded_annotation",
+    help = "Directory to write output files",
+  )
+)
+
+# Parse options
+opt <- parse_args(OptionParser(option_list = option_list))
+
+# Make output paths 
+expanded_fasta <- file.path(opt$output_dir, 
+                            "ensembl.GRCh38.annotation.expanded.fa")
+expanded_gtf <- file.path(opt$output_dir, 
+                          "ensembl.GRCh38.annotation.expanded.gtf")
+expanded_tx2gene <- file.path(opt$output_dir, 
+                              "ensembl.GRCh38.annotation.expanded.tx2gene.tsv")
+
+# Check for output directory 
+if (!dir.exists(opt$output_dir)) {
+  dir.create(opt$output_dir)
+}
+
+# extract GRanges object containing genomic coordinates of each 
+# annotated transcript and intron 
+gtf <- opt$gtf
+grl <- eisaR::getFeatureRanges(
+  gtf = gtf,
+  featureType = c("spliced", "intron"), 
+  intronType = "separate", 
+  flankLength = 90L, 
+  joinOverlappingIntrons = FALSE,
+  verbose = TRUE
+)
+#grl[4:6]
+
+# load in primary genome
+genome <- Biostrings::readDNAStringSet(opt$genome)
+head(genome)
+names(genome) <- sapply(strsplit(names(genome), " "), .subset, 1)
+
+# extract sequences for features of interest based on spliced and intronic
+# genomic regions defined above
+seqs <- GenomicFeatures::extractTranscriptSeqs(
+  x = genome, 
+  transcripts = grl
+)
+
+# write regions of interest to fasta file
+Biostrings::writeXStringSet(
+  seqs, filepath = expanded_fasta
+)
+
+# write the associated annotations to gtf 
+eisaR::exportToGtf(
+  grl, 
+  filepath = expanded_gtf
+)
+
+# create text file mapping transcript and intron identifiers to 
+# corresponding gene identifiers
+df <- eisaR::getTx2Gene(
```

Comment:
Can we use this to generate the Tx2Gene for the main index too, or do we want to generate _all_ of the transcript references with this script?

Reply:
So I was looking into this and theoretically yes we could grab the spliced and unspliced genomic regions and then generate the Tx2Gene mapping with `eisaR` for the main index. My concern with doing this, which is potentially minor, but in the previous script, you had grabbed the versioned transcript ID's. The way that the `eisaR::getTx2Gene` works is that it converts from a GenomicRanges object. Currently using the `eisaR::getFeatureRanges` function it doesn't look like I can grab versioned transcript ID's but only get spliced vs. unspliced. So I'm not sure if we would want that for our main index? Additionally, we would still have to use AnnotationHub to extract the mito genes for QC. 

So I'm thinking I will just modify the scripts to get the regions corresponding to unspliced transcripts rather than intronic regions and then merge `build_tx2gene_mito.R` and `make_pre_mrna_fasta.R` into one script that for now will be run outside of a docker container.

Reply:
I'm not sure we care about versioned transcript IDs, especially if we generate our own transcriptome file... If we can't get versioned transcripts for unspliced RNAs, then we probably don't want them if we are to do any comparisons? I mean, I don't know how comparable single-nucleus and single-cell will be, but it seems we might want to make an attempt? I have to think about it. 

Either way, we ought to be able to get the mitochrondrial gene list out of the gtf as well, so I think we could avoid AnnotationHub (which is something I think I would prefer).

Reply:
Yes totally makes sense. I went ahead and merged them into one and made it so we were generating all the needed files without using AnnotationHub. It should work and give us what we need for now.